README.md
# Readme

[![Build Status](https://travis-ci.org/torproject/sbws.svg?branch=master)](https://travis-ci.org/https://travis-ci.org/torproject/sbws)

Simple Bandwidth Scanner (called `sbws`) is a Tor bandwidth scanner that
produces bandwidth files to be used by Directory Authorities.

The scanner builds two hop circuits consisting of the relay being measured and
a fast exit. Over these circuits it measures bandwidth and store the results.

The generator read the measurement results, scales them using torflow's
scaling method and creates the bandwidth file.

**WARNING**: This software is intended to be run by researchers using a test
Tor network, such as chutney or shadow, or by the Tor bandwidth authorities
on the public Tor network.
Please do not run this software on the public Tor network unless you are one
of the Tor bandwidth authorities, to avoid creating unnecessary traffic.

**ADVICE**: It is recommended to read this documentation at
[Read the Docs](https://sbws.rtfd.io). In
[Github](http://github.com/torproject/sbws) some links won't be properly
rendered.
It can also be read after installing the Debian package ``sbws-doc`` in
``/usr/share/doc/sbws`` or after building it locally as explained in
``./docs/source/documenting.rst``.


Installing
------------

See [./INSTALL.rst](INSTALL.rst) (in local directory or GitHub) or
[INSTALL.html](INSTALL.html) (local build or Read the Docs).

Deploying and running
---------------------

See [./DEPLOY.rst](DEPLOY.rst) (in local directory or GitHub) or
[DEPLOY.html](DEPLOY.html) (local build or Read the Docs).

Contributing
--------------

See [./CONTRIBUTING.rst](CONTRIBUTING.rst) (in local directory or GitHub) or
[CONTRIBUTING.html](CONTRIBUTING.html) (local build or Read the Docs).

Changelog
--------------

See [./CHANGELOG.rst](CHANGELOG.rst) (in local directory or GitHub) or
[CHANGELOG.html](CHANGELOG.html)  (local build or Read the Docs).

Documentation
--------------

More extensive documentation can be found in the ``./docs`` directory,
and online at [sbws.rtfd.io](https://sbws.readthedocs.io).

## License

This work is in the public domain within the United States.

We waive copyright and related rights in the work worldwide through the
[CC0-1.0 license](https://creativecommons.org/publicdomain/zero/1.0).

You can find a copy of the CC0 Public Domain Dedication along with this
software in ./LICENSE.md

## Authors

See [./AUTHORS.md](AUTHORS.md) (in local directory or GitHub) or
[AUTHORS.html](AUTHORS.html) (local build or Read the Docs).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- Ensure de default scaling method is torflow, leave cli option but ignore it,
  update documentation reflecting this. (#28446).

### Fixed

- Make sbws round to 3 significant figures in torflow rounding mode, rather
  than rounding to the nearest 1000 kilobytes (#28442).
- Continue running with defaults when the config argument is provided
  but not found (#28500).
- Make sbws round to 2 significant figures by default. This implements part
  of proposal 276 (#28451).
- Stop trying to resolve destination domain locally and unify criteria to
  choose relays as exits (#28458, #28471). Bugfix 1.0.4

## [1.0.2] - 2018-11-10

### Fixed

- Update bandwidth file specification version in the `generator` (#28366).
- Use 5 "=" characters as terminator in the bandwidth files (#28379)

### Changed

- Include the headers about eligible relays in all the bandwidth files,
  not only in the ones that does not have enough eligible relays (#28365).

## [1.0.1] - 2018-11-01

### Changed

- Change default directories when sbws is run from a system service (#28268).

## [1.0.0] - 2018-10-29

**Important changes**:

- `generate` includes extra statistics header lines when the number of
  eligible relays to include is less than the 60% of the network.
  It does not include the relays' lines.
- Speed up `scanner` by disabling RTT measurements and waiting for
  measurement threads before prioritizing again the list of relays to measure.

### Fixed

- Update python minimal version in setup (#28043)
- Catch unhandled exception when we fail to resolve a domain name (#28141)
- Bandwidth filtered is the maximum between the bandwidth measurements and
  their mean, not the minimum (#28215)
- Stop measuring the same relay by two threads(#28061)

### Changed

- Move ``examples/`` to ``docs/`` (#28040)
- Number of results comparison and number of results away from each other are
  incorrect (#28041)
- Stop removing results that are not away from some other X secs (#28103)
- Use secs-away when provided instead of data_period (#28105)
- Disable measuring RTTs (#28159)
- Rename bandwidth file keyvalues (#28197)

## Added

- Write bw file only when the percentage of measured relays is bigger than 60%
  (#28062)
- When the percentage of measured relays is less than the 60%, do not include
  the relays in the bandwidth file and instead include some statistics in the
  header (#28076)
- When the percentage of measured relays is less than the 60% and it was more
  before, warn about it (#28155)
- When the difference between the total consensus bandwidth and the total
  in the bandwidth lines is larger than 50%, warn (#28216)
- Add documentation about how the bandwidth measurements are selected and
  scaled before writing them to the Bandwidth File (#27692)

## [0.8.0] - 2018-10-08

**Important changes**:

- Implement Torflow scaling/aggregation to be able to substitute Torflow with
  sbws without affecting the bandwidth files results.
- Change stem dependency to 1.7.0, which removes the need for `dependency_links``
- Update and cleanup documentation

### Added

- Add system physical requirements section to INSTALL (#26937)
- Warn when there is not enough disk space (#26937)
- Implement Torflow scaling (#27108)
- Create methods to easy graph generation and obtain statistics
  to compare with current torflow results.(#27688)
- Implement rounding bw in bandwidth files to 2 insignificant digits(#27337)
- Filter results in order to include relays in the bandwidth file that:(#27338)
  - have at least two measured bandwidths
  - the measured bandwidths are within 24 hours of each other
  - have at least two descriptor observed bandwidths
  - the descriptor observed bandwidths are within 24 hours of each other

### Fixed

- Broken environment variable in default sbws config. To use envvar $FOO, write
  $$FOO in the config.
- Stop using directory as argument in integration tests (#27342)
- Fix typo getting configuration option to allow logging to file (#27960)
- Set int type to new arguments that otherwise would be string (#27918)
- Stop printing arguments default values, since they are printed by default
  (#27916)
- Use dash instead of underscore in new cli argument names (#27917)

### Changed

- sbws install doc is confusing (#27341)
  - Include system and Python dependencies in `INSTALL`.
  - Include dependencies for docs and tests in `INSTALL`.
  - Point to `DEPLOY` to run sbws.
  - Remove obsolete sections in `INSTALL`
  - Simplify `DEPLOY`, reuse terms in the `glossary`.
  - Remove obsolete ``sbws init`` from `DEPLOY`.
  - Point to config documentation.
  - Add, unify and reuse terms in `glossary`.
- refactor v3bwfile (#27386): move scaling method inside class
- use custom ``install_command`` to test installation commands while
  ``dependency_links`` is needed until #26914 is fixed. (#27704)
- documentation cleanup (#27773)
  - split, merge, simplify, extend, reorganize sections and files
- generate scales as Torflow by default (#27976)
- Replace stem ``dependency_links`` by stem 1.7.0 (#27705). This also eliminates
  the need for custom ``install_command`` in tox.

## [0.7.0] - 2018-08-09

**Important changes**:

- `cleanup/stale_days` is renamed to `cleanup/data_files_compress_after_days`
- `cleanup/rotten_days` is renamed to `cleanup/data_files_delete_after_days`
- sbws now takes as an argument the path to a config file (which contains
  `sbws_home`) instead of `sbws_home` (which contains the path to a config
file)

### Added

- Log line on start up with sbws version, platform info, and library versions
(trac#26751)
- Manual pages (#26926)

### Fixed

- Stop deleting the latest.v3bw symlink. Instead, do an atomic rename.
  (#26740)

sbws/util/fs.py
"""Utils file system functions"""

import logging
import shutil

log = logging.getLogger(__name__)

DISK_SPACE_TEXT = """
Disk space requirements
-----------------------
v3bw files: the maximum space required is ~{mb_bw} MB, after {d_bw} days.
result files: the maximum space required is ~{mb_results} MB, after {d_r} days.
tor directory: the space required is ~{mb_tor} MB.
code and depenencies: the space required is ~{mb_code} MB
Total disk space required is: ~{mb_total} MB
"""


def sbws_required_disk_space(conf):
    """Disk space required by sbws files.
    Rough calculations.

    :param ConfigParser conf: sbws configuration
    :returns: int, size in MB
    """
    text_dict = {}
    # Number of relays per line average size in Bytes
    size_v3bw_file = 7500 * 220
    # default crontab configuration will run genenerate every hour
    num_v3bw_files_day = 24
    # ~1000 is the length of a line when the result is successfull
    # ~4550 is the number of lines of the biggest result file
    size_result_file = 4550 * 1000
    num_result_files_day = 1
    space_v3bw_files_day = size_v3bw_file * num_v3bw_files_day
    space_result_files_day = size_result_file * num_result_files_day
    size_compressed_files = 600 * 1024
    # default crontab configuration will run cleanup once a day
    # default cleanup configuration will compress v3bw files after 1 day
    # and delete them after 7 days
    v3bw_compress_after_days = conf.getint('cleanup',
                                           'v3bw_files_compress_after_days')
    v3bw_delete_after_days = conf.getint('cleanup',
                                         'v3bw_files_delete_after_days')
    v3bw_max_space_after_delete = \
        (space_v3bw_files_day * v3bw_compress_after_days) + \
        (size_compressed_files * num_v3bw_files_day * v3bw_delete_after_days)
    text_dict['mb_bw'] = round(v3bw_max_space_after_delete / 1000 ** 2)
    text_dict['d_bw'] = v3bw_delete_after_days
    # default crontab configuration will run cleanup once a day
    # default cleanup configuration will compress v3bw files after 1 day
    # and delete them after 7 days
    results_compress_after_days = conf.getint('cleanup',
                                              'data_files_compress_after_days')
    results_delete_after_days = conf.getint('cleanup',
                                            'data_files_delete_after_days')
    results_max_space_after_delete = \
        (space_result_files_day * results_compress_after_days) + \
        (size_compressed_files * num_v3bw_files_day *
         results_delete_after_days)
    text_dict['mb_results'] = round(results_max_space_after_delete / 1000 ** 2)
    text_dict['d_r'] = results_delete_after_days
    # not counted rotated files and assuming that when it is not rotated the
    # size will be aproximately 10MiB
    space_log_files = 0
    if conf.getboolean('logging', 'to_file'):
        size_log_file = conf.getint('logging', 'to_file_max_bytes')
        num_log_files = conf.getint('logging', 'to_file_num_backups')
        space_log_files = size_log_file * num_log_files
    text_dict['mb_log'] = space_log_files
    # roughly, size of a current tor dir
    size_tor_dir = 19828000
    text_dict['mb_tor'] = round(size_tor_dir / 1000 ** 2)
    # roughly, the size of this code and dependencies
    size_code_deps = 2097152
    text_dict['mb_code'] = round(size_code_deps / 1000 ** 2)
    # Multiply per 2, just in case
    size_total = (results_max_space_after_delete +
                  v3bw_max_space_after_delete + space_log_files +
                  size_tor_dir + size_code_deps) * 2
    text_dict['mb_total'] = round(size_total / 1000 ** 2)
    space_text = DISK_SPACE_TEXT.format(**text_dict)
    return space_text


def df(path):
    # Not being used, since it makes a disk space system call and some
    # systems might not allow it
    """Return space left on device where path is in MiB."""
    return round(shutil.disk_usage(path).free / (1024 ** 2))


def is_low_space(conf):
    # Not being used, since it makes a disk space system call and some
    # systems might not allow it
    """Warn and return True when the space left on the device is less than
    what is needed for sbws and False otherwise needs.
    """
    disk_required_mb = sbws_required_disk_space(conf)
    disk_avail_mb = df(conf.getpath('paths', 'sbws_home'))
    if disk_avail_mb < disk_required_mb:
        log.warn("The space left on the device (%s MiB) is less than "
                 "the minimum recommended to run sbws (%s MiB)."
                 "Run sbws cleanup to delete old sbws generated files.",
                 disk_avail_mb, disk_required_mb)
        return True
    return False
